File: 백준/Silver/15654. N과 M （5）/N과 M （5）.cc
#include <bits/stdc++.h>
using namespace std;

int n, m;
int input[9];
int arr[9];
bool vis[10001];


void func(int target) {
    if (target == m) {
        for (int i = 0; i < m; i++) cout << arr[i] << " ";
        cout << '\n';
        return;
    }
    for (int i = 0; i < n; i++) {
        if (vis[input[i]] == true) continue;
        arr[target] = input[i];
        vis[input[i]] = true;
        func(target + 1);
        vis[input[i]] = false;
    }
}

int main(void) {
    cin.tie(0)->sync_with_stdio(false);
    cin >> n >> m;
    for (int i = 0; i < n; i++) cin >> input[i];
    sort(input, input + n);
    func(0);
}

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.io.*;
import java.util.*;

public class Main {

    static int n;
    static int m;
    static int[] input = new int[10];
    static int[] arr = new int[10];
    static boolean[] isused = new boolean[10]; // 인덱스 저장
    static StringBuilder sb = new StringBuilder();

    public static void backtracking(int cur) {
        if (cur == m) {
            for (int i = 0; i < m; i++) {
                sb.append(arr[i]).append(" ");
            }
            sb.append("\n");
            return;
        }

        for (int i = 0; i < n; i++) {
            if (isused[i]) continue;
            isused[i] = true;
            arr[cur] = input[i];
            backtracking(cur + 1);
            isused[i] = false;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            input[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(input, 0, n);
        backtracking(0);
        System.out.print(sb);
    }
}
